src/qtism/data/storage/xml/Utils.php

```diff
@@ -353,11 +355,40 @@ public static function valueAsString($value, $encode = true): string
             return $value === true ? 'true' : 'false';
         }
         if ($encode) {
-            return htmlspecialchars((string)$value, ENT_XML1, 'UTF-8');
+            return self::escapeNonCharacterRange((string)$value);
         }
         return (string)$value;
     }
 
+    public static function escapeNonCharacterRange(string $string): string
+    {
+        // Define the XML escape sequences
+        $symbolMap = [
+            '"'  => '&quot;',
+            "'"  => '&apos;',
+            '&'  => '&amp;',
+            '<'  => '&lt;',
+            '>'  => '&gt;',
+            "\t" => '&#x9;',
+            "\r" => '&#xD;',
+            "\x08" => '&#xFFFD;',
+        ];
+
+        // generate replace map
+        for ($char = 0x20; $char <= 0xDF77; $char++) {
+            $symbolMap[mb_chr($char, 'UTF-8')] = sprintf('&#x%X;', $char);
+        }
+        for ($char = 0xE000; $char <= 0xFFFD; $char++) {
+            $symbolMap[mb_chr($char, 'UTF-8')] = sprintf('&#x%X;', $char);
+        }
+        for ($char = 0x10000; $char <= 0x10FFFF; $char++) {
+            $symbolMap[mb_chr($char, 'UTF-8')] = sprintf('&#x%X;', $char);
+        }
+        $string = strtr($string, $symbolMap);
+
+        return htmlspecialchars($string, ENT_QUOTES | ENT_XML1, 'UTF-8');
```

Comment:
This will double encode some of the earlier replaced characters. I suggest you rely on the manual replacement map only.